### PR TITLE
Save the number of damaged objects

### PR DESCRIPTION
**Pull request by AymericDu**
##### SUMMARY

* Save the number of damaged objects in the meta2 database
* Use the event (`container.state`) to update the account
* Increase the counter for each chunk not uploaded
* Decrease the counter for each deleted object with missing chunks
* Decrease the counter for each rebuilt chunk (chunk not uploaded)
* Recompute the counter of damaged objects

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

* Python API
* `account`
* `proxy`
* `meta2`

##### SDS VERSION

```
openio 4.4.0.0b1.dev36
```

### CODE REVIEW

File: tests/functional/container/test_container.py [hunk 1]
```diff
@@ -540,12 +540,15 @@ def flush_and_check(truncated=False, objects=0, usage=0):
         flush_and_check(truncated=True, objects=16, usage=16384)
         flush_and_check()
 
-    def _check_missing_chunks(self, expected_missing_chunks):
+    def _check_missing_chunks(self, expected_damaged_objects,
```
Comment (murlock) [inline tests/functional/container/test_container.py]: It should be nice to have test with `append` and `truncate` method operations on object and `flush` on container.

File: meta2v2/meta2_backend.c [hunk 2]
```diff
@@ -1298,11 +1356,37 @@ meta2_backend_insert_beans(struct meta2_backend_s *m2b,
...
+					GString *content_id = metautils_gba_to_hexgstr(
+							NULL, CHUNKS_get_content(bean->data));
+					if (!g_hash_table_add(content_ids,
+							content_id->str)) {
+						g_string_free(content_id, TRUE);
+						continue;
+					}
+					oio_url_set(content_url, OIOURL_CONTENTID,
+							content_id->str);
+					g_string_free(content_id, TRUE);
```
Comment (fvennetier) [inline meta2v2/meta2_backend.c]: At this point, `content_id->str` is in the table, and should not be freed.
I suggest using `g_string_free(content_id, FALSE);`.

File: meta2v2/meta2_backend.c [hunk 3]
```diff
@@ -1298,11 +1356,37 @@ meta2_backend_insert_beans(struct meta2_backend_s *m2b,
 			else
 				err = _db_insert_beans_list (sq3->db, beans);
 			if (!err) {
+				gint64 damaged_objects = m2db_get_damaged_objects(sq3);
 				gint64 missing_chunks = m2db_get_missing_chunks(sq3);
+
+				GHashTable *content_ids = g_hash_table_new_full(
+						g_str_hash, g_str_equal, NULL, NULL);
```
Comment (fvennetier) [inline meta2v2/meta2_backend.c]: The key free function (3rd parameter) should be set to `g_free`. See comment below.

File: tests/functional/container/test_container.py [hunk 4]
```diff
@@ -434,6 +434,36 @@ def _create_content(self, name, missing_chunks=0):
...
+        resp = self.request('POST', self.url_content('create'), params=params,
+                            headers=headers, data=json.dumps(chunks))
+        self.assertEqual(204, resp.status)
+
+    def _truncate_content(self, name):
+        params = self.param_content(self.ref, name)
+        params['size'] = 1048576
+        resp = self.request('POST', self.url_content('truncate'),
+                            params=params)
+        print resp.data
```
Comment (murlock) [inline tests/functional/container/test_container.py]: Useless print ?

File: meta2v2/meta2_backend.c [hunk 5]
```diff
@@ -840,19 +896,43 @@ _update_missing_chunks(struct meta2_backend_s *m2b, struct sqlx_sqlite3_s *sq3,
...
 		content_missing_chunks = 0;
 		m2v2_sort_content(deleted_beans->data, &sorted);
+		if (!sorted->header) {
+			if (!header) {
+				// "truncate" and "update" doesn't delete the header
+				err = m2db_get_alias(sq3, url, M2V2_FLAG_NOPROPS,
+						retrieve_header, NULL);
+			}
+			if (err || !header) {
+                if (err) {
```
Comment (murlock) [inline meta2v2/meta2_backend.c]: Tab or Space strikes again ?

**Pull request merged**